### PR TITLE
Exercise 4

### PR DESCRIPTION
**Pull request by Psyhich**
Implemented #8

### CODE REVIEW

File: Exercise/include/my_string.h [hunk 1]
```diff
@@ -89,9 +90,28 @@ namespace MyStructs
 			ToLowerCase(0, size() - 1);
 		}
 
+		CMyString Reverse() const noexcept;
+
+		std::optional<int> ToInt() const noexcept;
+		static CMyString FromInt(int iToConvert) noexcept;
```
Comment (VictorApopiy) [inline Exercise/include/my_string.h:97]: const?

File: Exercise/include/my_string.h [hunk 2]
```diff
@@ -111,6 +131,16 @@ namespace MyStructs
 			return nPosition > size() || nCountOfChars == 0 || 
 				nCountOfChars > size();
 		}
+		/*
+		* Can only be used in compare functions, don't use it raw
+		* */
+		static bool CheckEquality(const char* cszFirstString, const char *cszSecondString, size_t nLengthToCheck);
+		static bool CheckEqualityCaseInsensitive(const char* cszFirstString, const char *cszSecondString, size_t nLengthToCheck);
```
Comment (VictorApopiy) [inline Exercise/include/my_string.h:143]: const?

File: Exercise/src/my_string.cpp [hunk 3]
```diff
@@ -1,8 +1,53 @@
...
+#include <cmath>
+#include <limits>
 
 #include "my_string.h"
+
+char MyStructs::CMyString::ToLowerCase(char chCharToTranslate)
+{
+	if(chCharToTranslate >= 'A' && chCharToTranslate < 'Z')
+	{
+		return chCharToTranslate + 32;
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: Define this number please

File: Exercise/src/my_string.cpp [hunk 4]
```diff
@@ -1,8 +1,53 @@
...
+}
+
+char MyStructs::CMyString::ToUpperCase(char chCharToTranslate)
+{
+	if(chCharToTranslate >= 'a' && chCharToTranslate <= 'z')
+	{
+		return chCharToTranslate - 32;
+	}
+	return chCharToTranslate;
+}
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: Please create the similar function that takes full string as an argument
Same form lowercase

File: Exercise/src/my_string.cpp [hunk 5]
```diff
@@ -397,13 +447,373 @@ bool MyStructs::CMyString::operator==(const CMyString& cStringToCompare) const
...
+	{
+		printf("Given range is invalid!!\n");
+		return newString;
+	}
+
+	if(nPos + nLength <= size())
+	{
+		--nLength; // Minusing one to exclude terminant from cszStrToInput
+		newString.m_nSize = size();
+	} else
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:714]: add new line

File: Exercise/src/my_string.cpp [hunk 6]
```diff
@@ -397,13 +447,373 @@ bool MyStructs::CMyString::operator==(const CMyString& cStringToCompare) const
...
+	// Subtracting to make it as length value, not the position(because we start from 0)
+	if(nStartPos == 0)
+	{
+		--nLength;
+	}
+	
+	if(bIsCaseSensitive)
+	{
+		return CheckEquality(data() + nStartPos, cszStringToCompare, nLength);
+	} else
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:669]: \n

Comment (Psyhich) [reply Exercise/src/my_string.cpp:669]: Fixed

File: Exercise/src/my_string.cpp (hunk 5 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:714]: Fixed

File: Exercise/src/my_string.cpp (hunk 4 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: There are already method that does to lowercase or uppercase transformation, these are static methods to shorten the repetition of same functionality

File: Exercise/src/my_string.cpp (hunk 3 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Defined

File: Exercise/include/my_string.h (hunk 2 above)
Comment (Psyhich) [reply Exercise/include/my_string.h:143]: I cannot make static functions a constant ones

File: Exercise/include/my_string.h (hunk 1 above)
Comment (Psyhich) [reply Exercise/include/my_string.h:97]: Cannot make const static functions

**Review by VictorApopiy — APPROVED** [review]

**Pull request merged**